fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushDownAggThroughJoinByFk.java

```diff
@@ -0,0 +1,146 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.rewrite;
+
+import org.apache.doris.nereids.properties.DataTrait;
+import org.apache.doris.nereids.properties.FuncDeps;
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.algebra.Project;
+import org.apache.doris.nereids.trees.plans.logical.LogicalAggregate;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.util.JoinUtils;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import org.apache.thrift.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Push down agg through join with foreign key:
+ *    Agg(group by fk/pk)
+ *     |
+ *   Join(pk = fk)
+ *   /  \
+ *  pk  fk
+ *  ======>
+ *   Join(pk = fk)
+ *   /     \
+ *  |  Agg(group by fk)
+ *  |      |
+ *  pk    fk
+ */
+public class PushDownAggThroughJoinByFk implements RewriteRuleFactory {
+    @Override
+    public List<Rule> buildRules() {
+        return ImmutableList.of(
+                logicalAggregate(
+                        innerLogicalJoin()
+                                .when(j -> j.getJoinType().isInnerJoin()
+                                        && !j.isMarkJoin()
+                                        && j.getOtherJoinConjuncts().isEmpty()))
+                        .when(agg -> agg.getGroupByExpressions().stream().allMatch(Slot.class::isInstance)
+                                    && agg.getOutputExpressions().stream().allMatch(Slot.class::isInstance))
+                        .thenApply(ctx -> pushAgg(ctx.root, ctx.root.child()))
+                        .toRule(RuleType.PUSH_DOWN_AGG_THROUGH_FK_JOIN),
+                logicalAggregate(
+                        logicalProject(
+                                innerLogicalJoin()
+                                        .when(j -> j.getJoinType().isInnerJoin()
+                                                && !j.isMarkJoin()
+                                                && j.getOtherJoinConjuncts().isEmpty()))
+                                .when(Project::isAllSlots))
+                        .when(agg -> agg.getGroupByExpressions().stream().allMatch(Slot.class::isInstance))
+                        .thenApply(ctx -> pushAgg(ctx.root, ctx.root.child().child()))
+                        .toRule(RuleType.PUSH_DOWN_AGG_THROUGH_FK_JOIN)
+        );
+    }
+
+    private @Nullable Plan pushAgg(LogicalAggregate<?> agg, LogicalJoin<?, ?> join) {
+        Plan foreign = tryExtractForeign(join);
```

Comment:
better to extract this part of logic, i.e, pk-fk join checking, at the entry point of this rule?
current checking seems not clear and easy understanding.

Reply:
if put in the entry point of the rule, we should extract the foreign table twice in pushAgg and entry point.